[CMS-834] Hook into graphql to emit the right surrogate keys.

kporras07 opened the pull request.

jspellman814 commented on inc/class-emitter.php:
```diff
@@ -371,4 +378,55 @@ function( $a, $b ) {
+	/**
+	 * Send additional headers to graphql response.
+	 */
+	public static function filter_graphql_response_headers_to_send($headers) {
```
```suggestion
	public static function filter_graphql_response_headers_to_send( $headers ) {
```

jspellman814 commented on inc/class-emitter.php:
```diff
@@ -371,4 +378,55 @@ function( $a, $b ) {
+	 * Inspect the model and get the right surrogate keys.
+	 */
+	public static function filter_graphql_dataloader_get_model( $model ) {
+		$reflect = new \ReflectionClass($model);
```
```suggestion
		$reflect = new \ReflectionClass( $model );
```

jspellman814 commented on inc/class-emitter.php:
```diff
@@ -371,4 +378,55 @@ function( $a, $b ) {
+		$class_short_name = $reflect->getShortName();
+		$surrogate_key_prefix = strtolower($class_short_name);
+		if ( isset( $model->id )) {
+			if (!empty($model->databaseId)) {
```
```suggestion
			if ( ! empty( $model->databaseId ) ) {
```

jspellman814 commented on inc/class-emitter.php:
```diff
@@ -371,4 +378,55 @@ function( $a, $b ) {
+			if (!empty($model->databaseId)) {
+				self::get_instance()->graphql_surrogate_keys[] = $surrogate_key_prefix . '-' . $model->databaseId;
+			}
+			if (!empty($model->slug)) {
```
```suggestion
			if ( ! empty( $model->slug ) ) {
```

jspellman814 commented on inc/class-emitter.php:
```diff
@@ -371,4 +378,55 @@ function( $a, $b ) {
+	public static function filter_graphql_dataloader_get_model( $model ) {
+		$reflect = new \ReflectionClass($model);
+		$class_short_name = $reflect->getShortName();
+		$surrogate_key_prefix = strtolower($class_short_name);
```
```suggestion
		$surrogate_key_prefix = strtolower( $class_short_name );
```

jspellman814 reviewed (changes requested):
~~Few small formatting changes to maintain consistency with the rest of the plugin.~~ Linting picked up on these 😄  Otherwise looking good to me.

jspellman814 reviewed (approved):
LGTM! Let's hold on merging until we hear from Decoupled, though.

backlineint reviewed (changes requested):
I see the surrogate key header on wp-graphql responses, which is great. And it seems to handle cases with multiple results well. But I'm seeing more keys than I would expect. 

Take an individual piece of page content - if I look at the header when viewing the page in WordPress, I see the following keys:

post-2 post-user-1 single

If I use a graphql query for that page:

```
query NewQuery {
  page(id: "cG9zdDoy") {
        id
        content
  }
}
```

I see the following keys:

post-2 post-sample-page user-1 user-decoupledadmin

That more than doubles the header, with some info that I don't thing Fastly would use. Not a huge deal for one page, but I wonder if it could be an issue for a query that returns something like 50 results. From the plugin docs:

> There is a limit to the number of surrogate keys in a response, so we’ve optimized them based on a user’s expectation of a normal WordPress site

Also, I wonder about leaking the username here.

So I'd be interested to know how much effort would be required to remove the keys that aren't included for the standard WP response. That seems more consistent and efficient. That said, the current implementation is really close and would likely work in the majority of cases.

kporras07 commented:
@backlineint Quick question:

Currently, we have:
- Page: post-2 post-user-1 single
- GraphQL: post-2 post-sample-page user-1 user-decoupledadmin

I can easily make GraphQL to look like this: post-2 user-1

1) Should I also change “user-1”  to “post-user-1"?
2) Do you really need the "single" key to be outputted via graphql? (That one will likely take some more effort)

kporras07 commented:
I went ahead and removed the slug keys and added post-user-* keys for posts (lmk if we need this for any other "entity type"). Also, I added some code that will remove a "user-*" key if there is a corresponding "post-user-*" key.

Currently open questions:
1) Should we add any other keys similar to "post-user-*"?
2) Should we add a "single" key under certain conditions?

kporras07 commented:
@backlineint updated to match recent Slack conversation.

Example output:

```
post-9 post-1 user-1 graphql-collection
```

kporras07 commented:
It was a request from @backlineint when I asked him via Slack about this

backlineint reviewed (approved):
Did some testing with the latest - looks right to me. Thanks for the quick revisions!
